File: examples/java/BoltUtilDemo.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Instant;
import java.util.Locale;

public class BoltUtilDemo {
    private static final String API_BASE = env("BOLT_API_BASE", "https://api.boltutil.com");
    private static final String API_KEY = env("BOLT_API_KEY", "YOUR_API_KEY");
    private static final String WEBHOOK_SECRET = env("BOLT_WEBHOOK_SECRET", "YOUR_WEBHOOK_SECRET");
    private static final String NOTIFY_URL = env("BOLT_NOTIFY_URL", "https://merchant.example.com/webhooks/boltutil");
    private static final String RETURN_URL = env("BOLT_RETURN_URL", "https://merchant.example.com/orders/return");

    private static String env(String key, String fallback) {
        String value = System.getenv(key);
        return value == null || value.isBlank() ? fallback : value;
    }

    private static String hmacSha256Hex(String payload, String secret) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
        byte[] bytes = mac.doFinal(payload.getBytes(StandardCharsets.UTF_8));

        StringBuilder hex = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            hex.append(String.format("%02x", b));
        }
        return hex.toString();
    }

    private static String signBody(String rawBody, String timestamp) throws Exception {
        return hmacSha256Hex(timestamp + "." + rawBody, WEBHOOK_SECRET);
    }

    private static void createOrder(String network) throws Exception {
        String externalOrderId = "INV-" + Instant.now().toEpochMilli();
        String rawBody = "{"
                + "\"amount\":25.0,"
                + "\"externalOrderId\":\"" + externalOrderId + "\","
                + "\"network\":\"" + network + "\","
                + "\"currency\":\"USDT\","
                + "\"orderDesc\":\"BoltUtil demo order\","
                + "\"notifyUrl\":\"" + NOTIFY_URL + "\","
                + "\"returnUrl\":\"" + RETURN_URL + "?order=" + externalOrderId + "\","
                + "\"expiredMinutes\":30,"
                + "\"metadata\":{\"customerId\":\"CUS_1001\",\"source\":\"java-demo\"}"
                + "}";

        String timestamp = String.valueOf(Instant.now().toEpochMilli());
        String signature = signBody(rawBody, timestamp);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(API_BASE + "/api/v1/order/create"))
                .header("Content-Type", "application/json")
                .header("X-Bolt-Key", API_KEY)
                .header("X-Bolt-Timestamp", timestamp)
                .header("X-Bolt-Signature", signature)
                .POST(HttpRequest.BodyPublishers.ofString(rawBody, StandardCharsets.UTF_8))
                .build();

        HttpResponse<String> response = HttpClient.newHttpClient()
                .send(request, HttpResponse.BodyHandlers.ofString());

        System.out.println("HTTP " + response.statusCode());
        System.out.println(response.body());
    }

    private static boolean verifyWebhook(String rawBody, String timestamp, String signature) throws Exception {
        if (timestamp == null || timestamp.isBlank() || signature == null || signature.isBlank()) {
            return false;
        }

        long ts = Long.parseLong(timestamp);
        if (Math.abs(Instant.now().toEpochMilli() - ts) > 5 * 60 * 1000) {
            return false;
        }

        String expected = signBody(rawBody, timestamp).toLowerCase(Locale.ROOT);
        String received = signature.toLowerCase(Locale.ROOT);
        return MessageDigest.isEqual(
                expected.getBytes(StandardCharsets.UTF_8),
                received.getBytes(StandardCharsets.UTF_8)
        );
    }

    private static void startWebhookServer(int port) throws IOException {
        HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);
        server.createContext("/webhooks/boltutil", BoltUtilDemo::handleWebhook);
        server.start();
        System.out.println("BoltUtil webhook demo listening on http://localhost:" + port + "/webhooks/boltutil");
    }

    private static void handleWebhook(HttpExchange exchange) throws IOException {
        if (!"POST".equalsIgnoreCase(exchange.getRequestMethod())) {
            sendJson(exchange, 404, "{\"error\":\"not_found\"}");
            return;
        }

        String rawBody = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
        String timestamp = exchange.getRequestHeaders().getFirst("X-Bolt-Webhook-Timestamp");
        String signature = exchange.getRequestHeaders().getFirst("X-Bolt-Webhook-Signature");

        try {
            if (!verifyWebhook(rawBody, timestamp, signature)) {
                sendJson(exchange, 401, "{\"error\":\"invalid_signature\"}");
                return;
            }
        } catch (Exception e) {
            sendJson(exchange, 401, "{\"error\":\"invalid_signature\"}");
            return;
        }

        System.out.println("Verified BoltUtil webhook: " + rawBody);
        // TODO: update your local order by externalOrderId from the JSON body.
        // Only fulfill when status is CONFIRMED or COMPLETED, according to your business rules.

        sendJson(exchange, 200, "{\"status\":\"SUCCESS\"}");
    }

    private static void sendJson(HttpExchange exchange, int statusCode, String body) throws IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(statusCode, bytes.length);
        try (OutputStream output = exchange.getResponseBody()) {
            output.write(bytes);
        }
    }

    public static void main(String[] args) throws Exception {
        String command = args.length > 0 ? args[0] : "";
        if ("create".equals(command)) {
            createOrder(args.length > 1 ? args[1] : "TRC20");
        } else if ("webhook".equals(command)) {
            startWebhookServer(Integer.parseInt(env("PORT", "3000")));
        } else {
            System.out.println("Usage:");
            System.out.println("  javac BoltUtilDemo.java && java BoltUtilDemo create TRC20");
            System.out.println("  javac BoltUtilDemo.java && java BoltUtilDemo create ERC20");
            System.out.println("  javac BoltUtilDemo.java && java BoltUtilDemo create BEP20");
            System.out.println("  javac BoltUtilDemo.java && java BoltUtilDemo webhook");
        }
    }
}

File: examples/csharp/BoltUtilDemo.cs
using System;
using System.Net;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

public static class BoltUtilDemo
{
    private static readonly string ApiBase = Env("BOLT_API_BASE", "https://api.boltutil.com");
    private static readonly string ApiKey = Env("BOLT_API_KEY", "YOUR_API_KEY");
    private static readonly string WebhookSecret = Env("BOLT_WEBHOOK_SECRET", "YOUR_WEBHOOK_SECRET");
    private static readonly string NotifyUrl = Env("BOLT_NOTIFY_URL", "https://merchant.example.com/webhooks/boltutil");
    private static readonly string ReturnUrl = Env("BOLT_RETURN_URL", "https://merchant.example.com/orders/return");

    private static string Env(string key, string fallback)
    {
        var value = Environment.GetEnvironmentVariable(key);
        return string.IsNullOrWhiteSpace(value) ? fallback : value;
    }

    private static string HmacSha256Hex(string payload, string secret)
    {
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(payload));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    private static string SignBody(string rawBody, string timestamp)
    {
        return HmacSha256Hex($"{timestamp}.{rawBody}", WebhookSecret);
    }

    private static async Task CreateOrder(string network)
    {
        var externalOrderId = $"INV-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}";
        var rawBody = "{"
            + "\"amount\":25.0,"
            + $"\"externalOrderId\":\"{externalOrderId}\","
            + $"\"network\":\"{network}\","
            + "\"currency\":\"USDT\","
            + "\"orderDesc\":\"BoltUtil demo order\","
            + $"\"notifyUrl\":\"{NotifyUrl}\","
            + $"\"returnUrl\":\"{ReturnUrl}?order={externalOrderId}\","
            + "\"expiredMinutes\":30,"
            + "\"metadata\":{\"customerId\":\"CUS_1001\",\"source\":\"csharp-demo\"}"
            + "}";

        var timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString();
        var signature = SignBody(rawBody, timestamp);

        using var client = new HttpClient();
        using var request = new HttpRequestMessage(HttpMethod.Post, $"{ApiBase}/api/v1/order/create");
        request.Content = new StringContent(rawBody, Encoding.UTF8, "application/json");
        request.Headers.Add("X-Bolt-Key", ApiKey);
        request.Headers.Add("X-Bolt-Timestamp", timestamp);
        request.Headers.Add("X-Bolt-Signature", signature);

        using var response = await client.SendAsync(request);
        Console.WriteLine($"HTTP {(int)response.StatusCode}");
        Console.WriteLine(await response.Content.ReadAsStringAsync());
    }

    private static bool VerifyWebhook(string rawBody, string timestamp, string signature)
    {
        if (string.IsNullOrWhiteSpace(timestamp) || string.IsNullOrWhiteSpace(signature))
        {
            return false;
        }

        if (!long.TryParse(timestamp, out var ts))
        {
            return false;
        }

        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        if (Math.Abs(now - ts) > 5 * 60 * 1000)
        {
            return false;
        }

        var expected = SignBody(rawBody, timestamp).ToLowerInvariant();
        var received = signature.ToLowerInvariant();
        return CryptographicOperations.FixedTimeEquals(
            Encoding.UTF8.GetBytes(expected),
            Encoding.UTF8.GetBytes(received)
        );
    }

    private static async Task StartWebhookServer(int port)
    {
        using var listener = new HttpListener();
        listener.Prefixes.Add($"http://0.0.0.0:{port}/webhooks/boltutil/");
        listener.Start();
        Console.WriteLine($"BoltUtil webhook demo listening on http://localhost:{port}/webhooks/boltutil");

        while (true)
        {
            var context = await listener.GetContextAsync();
            _ = Task.Run(async () => await HandleWebhook(context));
        }
    }

    private static async Task HandleWebhook(HttpListenerContext context)
    {
        if (context.Request.HttpMethod != "POST")
        {
            await SendJson(context, 404, "{\"error\":\"not_found\"}");
            return;
        }

        using var reader = new System.IO.StreamReader(context.Request.InputStream, Encoding.UTF8);
        var rawBody = await reader.ReadToEndAsync();
        var timestamp = context.Request.Headers["X-Bolt-Webhook-Timestamp"] ?? "";
        var signature = context.Request.Headers["X-Bolt-Webhook-Signature"] ?? "";

        if (!VerifyWebhook(rawBody, timestamp, signature))
        {
            await SendJson(context, 401, "{\"error\":\"invalid_signature\"}");
            return;
        }

        Console.WriteLine("Verified BoltUtil webhook:");
        Console.WriteLine(rawBody);
        // TODO: update your local order by externalOrderId from the JSON body.
        // Only fulfill when status is CONFIRMED or COMPLETED, according to your business rules.

        await SendJson(context, 200, "{\"status\":\"SUCCESS\"}");
    }

    private static async Task SendJson(HttpListenerContext context, int statusCode, string body)
    {
        var bytes = Encoding.UTF8.GetBytes(body);
        context.Response.StatusCode = statusCode;
        context.Response.ContentType = "application/json";
        context.Response.ContentLength64 = bytes.Length;
        await context.Response.OutputStream.WriteAsync(bytes, 0, bytes.Length);
        context.Response.Close();
    }

    public static async Task Main(string[] args)
    {
        var command = args.Length > 0 ? args[0] : "";
        if (command == "create")
        {
            await CreateOrder(args.Length > 1 ? args[1] : "TRC20");
        }
        else if (command == "webhook")
        {
            await StartWebhookServer(int.Parse(Env("PORT", "3000")));
        }
        else
        {
            Console.WriteLine("Usage:");
            Console.WriteLine("  dotnet run -- create TRC20");
            Console.WriteLine("  dotnet run -- create ERC20");
            Console.WriteLine("  dotnet run -- create BEP20");
            Console.WriteLine("  dotnet run -- create POLYGON");
            Console.WriteLine("  dotnet run -- create SOLANA");
            Console.WriteLine("  dotnet run -- webhook");
        }
    }
}
